lang/java/reef-webserver/src/main/java/org/apache/reef/webserver/HttpServerImpl.java

```diff
@@ -70,39 +72,56 @@
                  @Parameter(MaxPortNumber.class) final int maxPortNumber,
                  @Parameter(MinPortNumber.class) final int minPortNumber,
                  @Parameter(MaxRetryAttempts.class) final int maxRetryAttempts,
+                 final TcpPortProvider tcpPortProvider,
                  final LoggingScopeFactory loggingScopeFactory) throws Exception {
 
     this.loggingScopeFactory = loggingScopeFactory;
+    this.jettyHandler = jettyHandler;
+    int availablePort = portNumber;
+    Server srv = null;
+
     try (final LoggingScope ls = this.loggingScopeFactory.httpServer()) {
-      this.jettyHandler = jettyHandler;
-      int availablePort = portNumber;
-      Server srv = null;
-      boolean found = false;
-      for (int attempt = 0; attempt < maxRetryAttempts; ++attempt) {
-        if (attempt > 0) {
-          availablePort = getNextPort(maxPortNumber, minPortNumber);
+      if (portNumber != Integer.parseInt(PortNumber.DEFAULT_VALUE)) {
+        // legacy path to be removed
+        for (int attempt = 0; attempt < maxRetryAttempts && srv  == null; ++attempt) {
+          if (attempt > 0) {
+            // first attempt should be portNumber passed in
+            availablePort = getNextPort(maxPortNumber, minPortNumber);
+          }
+          srv = tryPort(availablePort);
         }
-        srv = new Server(availablePort);
-        try {
-          srv.start();
-          found = true;
-          break;
-        } catch (final BindException ex) {
-          LOG.log(Level.FINEST, "Cannot use port: {0}. Will try another", availablePort);
+      } else {
+        // new TcpPortProvider path
+        Iterator<Integer> ports = tcpPortProvider.iterator();
```

Comment:
This could be captured in a foreach loop:

```java
for(final Integer portToTry : tcpPortProvider){
  srv = tryPort(portToTry);
  if(null != srv){
    break;
  }
}

Not sure whether this is nicer, though.
```